### PR TITLE
Better separation of customer code vs. FF code in schema upgrade demo

### PR DESCRIPTION
**Pull request by ChumpChief**
This change does a few things with the goal of better isolating the code written by the customer from code provided by FF:

1. Reorganizes some of the src directory structure, to better separate the "stuff the customer writes" from "stuff FF will provide".
2. Provides direct access to the `IMigrationTool` on the `IMigratableModel` for the `Migrator` to use rather than duplicating the interface on the `IMigratableModel` -- since the customer is responsible for implementing the `IMigratableModel` but the `MigrationTool` is something FF would provide, this makes it easier for the customer (see the appModel.ts files).
3. Gets a little more strict about readonly on interfaces, `import type`, interface documentation, avoid `enum`, etc.
4. A few naming tweaks to be more consistent about "container" vs. "app" vs. "model"

### CODE REVIEW

File: examples/hosts/app-integration/schema-upgrade/src/migrationTool/migrationTool.ts [hunk 1]
```diff
@@ -45,15 +44,23 @@ export class MigrationTool extends DataObject implements IMigrationTool {
...
-        return this.crc.read(newContainerIdKey) !== undefined;
+    public get migrationState() {
+        if (this.newContainerId !== undefined) {
+            return "migrated";
+        } else if (this.acceptedVersion !== undefined) {
+            return "migrating";
+        } else if (this.proposedVersion !== undefined) {
+            return "stopping";
+        } else {
+            return "collaborating";
```
Comment (scottn12) [inline examples/hosts/app-integration/schema-upgrade/src/migrationTool/migrationTool.ts:55]: Maybe add a note for what `"collaborating"` means (the other states feel more self-explanatory)

File: examples/hosts/app-integration/schema-upgrade/src/view/appView.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+        model.migrationTool.migrationState !== "collaborating",
+    );
+
+    useEffect(() => {
+        const migrationStateChangedHandler = () => {
+            setDisableInput(model.migrationTool.migrationState !== "collaborating");
+        };
+        model.migrationTool.on("stopping", migrationStateChangedHandler);
+        model.migrationTool.on("migrating", migrationStateChangedHandler);
+        model.migrationTool.on("migrated", migrationStateChangedHandler);
```
Comment (scottn12) [inline examples/hosts/app-integration/schema-upgrade/src/view/appView.tsx:35]: Thoughts on having `MigrationTool` have a single event (maybe called `"stateChange"` for example) and pass the new state value in the callback function? This way we could avoid having three listeners

**Review by scottn12 — APPROVED** [review]
Looks good! 😄

File: examples/hosts/app-integration/schema-upgrade/src/migrationTool/migrationTool.ts (hunk 1 above)
Comment (ChumpChief) [reply examples/hosts/app-integration/schema-upgrade/src/migrationTool/migrationTool.ts:55]: Added documentation for MigrationState in latest

File: examples/hosts/app-integration/schema-upgrade/src/view/appView.tsx (hunk 2 above)
Comment (ChumpChief) [reply examples/hosts/app-integration/schema-upgrade/src/view/appView.tsx:35]: Maybe - `Migrator` is the other consumer of the events and it cares a bit more about the specific states.  Also, if I trust the state machine enough then I could probably just listen for "stopping" here.  I'll think about this a bit more.

**Pull request merged**